core/pipeline-driver-queue/lib/service/preferred-jobs.js

```diff
@@ -0,0 +1,82 @@
+const queuePosition = require('@hkube/consts').queuePositions;
+const queueRunner = require('../queue-runner');
+const validator = require('../validation');
+const InvalidDataError = require('../errors/InvalidDataError');
+class PreferredJobs {
+    getPreferredJobsList() {
+        return queueRunner.preferredQueue.queue;
+    }
+
+    async deletePreferredJobs(jobIds) {
+        let deleted = 0;
+        const deletedJobs = jobIds.map(jobId => {
+            const deletedArr = queueRunner.preferredQueue.dequeue({ jobId });
+            if (deletedArr.length > 0) {
+                queueRunner.queue.enqueue(deletedArr[0]);
+                deleted += deletedArr.length;
+            }
+            return deletedArr;
+        });
+        if (deleted === 0) {
+            throw new InvalidDataError('JobIds do not exist in preferred queue');
+        }
+        return deletedJobs;
+    }
+
+    query(job, tag, pipeline, jobId) {
+        if (jobId) {
+            return job.jobId === jobId;
+        }
+        if (pipeline) {
+            return job.pipeline === pipeline;
+        }
+        if (tag) {
+            return job.tags.includes(tag);
+        }
+        return false;
+    }
+
+    addPreferredJobs({ addedJobs }) {
+        validator.preference.validatePreferenceRequest({ addedJobs });
+        const { ids, position, query } = addedJobs;
+        const { tag, pipeline, jobId } = query || {};
+        let index;
+        if (position === queuePosition.BEFORE) {
```

Comment:
This code is a little cumbersome.
Try to split to functions and add some comments